client/src/cmdlfem4x50.c

```diff
@@ -527,225 +633,684 @@ int em4x50_read(em4x50_data_t *etd, em4x50_word_t *out, bool verbose) {
+    } else {
+        PrintAndLogEx(FAILED, "Resetting password " _RED_("failed"));
         return PM3_ESOFT;
     }
 
+    // from now on new password 0x0
+    etd.password1 = 0x0;
+    
+    // clear data (words 1 to 31)
+    for (int i = 1; i < EM4X50_DEVICE_SERIAL; i++) {
+
+        // no login necessary for blocks 3 to 31
+        etd.pwd_given = (i <= EM4X50_CONTROL);
+        
+        PrintAndLogEx(INPLACE, "Wiping block %i", i);
+
+        etd.addresses = i << 8 | i;
+        clearCommandBuffer();
+        SendCommandNG(CMD_LF_EM4X50_WRITE, (uint8_t *)&etd, sizeof(etd));
+        if (!WaitForResponseTimeout(CMD_LF_EM4X50_WRITE, &resp, TIMEOUT)) {
+            PrintAndLogEx(WARNING, "Timeout while waiting for reply.");
+            return PM3_ETIMEOUT;
+        }
+
+        isOK = resp.status;
+        if (isOK != PM3_SUCCESS) {
+            PrintAndLogEx(NORMAL, "");
+            PrintAndLogEx(FAILED, "Wiping data " _RED_("failed"));
+            return PM3_ESOFT;
+        }
+    }
+
+    PrintAndLogEx(NORMAL, "");
+    PrintAndLogEx(SUCCESS, "Wiping data " _GREEN_("ok"));
+    
+    PrintAndLogEx(INFO, "Done");
+
     return PM3_SUCCESS;
 }
+
+int CmdEM4x50Restore(const char *Cmd) {
+
+    int uidLen = 0, fnLen = 0, pwdLen = 0, status = 0;
+    uint8_t pwd[4] = {0x0}, uid[4] = {0x0};
+    uint8_t data[DUMP_FILESIZE] = {0x0};
+    size_t bytes_read = 0;
+    char filename[FILE_PATH_SIZE] = {0};
+    em4x50_data_t etd = {.pwd_given = false};
+    PacketResponseNG resp;
+
+    CLIParserContext *ctx;
+    CLIParserInit(&ctx, "lf em 4x50_restore",
+                  "Restores data from dumpfile onto a Em4x50 tag.",
+                  "lf em 4x50_restore -u 1b5aff5c -> uses lf-4x50-1B5AFF5C-dump.bin\n"
+                  "lf em 4x50_restore -f mydump.eml -> uses mydump.eml\n"
+                  "lf em 4x50_restore -u 1b5aff5c -p 12345678 -> \n"
+                  "lf em 4x50_restore -f mydump.eml -p 12345678 -> \n"
+                 );
+
+    void *argtable[] = {
+        arg_param_begin,
+        arg_str0("u", "uid", "<uid>", "uid, hex, 4 bytes, msb, restore from lf-4x50-<uid>-dump.bin"),
+        arg_str0("f", "filename", "<filename>", "dump filename (bin/eml/json)"),
+        arg_str0("p", "passsword", "<password>", "password, hex, 4 bytes, lsb"),
+        arg_param_end
+    };
+
+    CLIExecWithReturn(ctx, Cmd, argtable, true);
+    
+    CLIGetHexWithReturn(ctx, 1, uid, &uidLen);
+    CLIParamStrToBuf(arg_get_str(ctx, 2),
+                     (uint8_t *)filename,
+                     FILE_PATH_SIZE,
+                     &fnLen
+                     );
+    CLIGetHexWithReturn(ctx, 3, pwd, &pwdLen);
+    
+    if ((uidLen && fnLen) || (!uidLen && !fnLen)) {
+        PrintAndLogEx(FAILED, "either use option 'u' or option 'f'");
+        return PM3_EINVARG;
+    }
+    
+    if (uidLen) {
+        snprintf(filename, FILE_PATH_SIZE, "./lf-4x50-%02x%02x%02x%02x-dump.bin",
```

Comment:
I think we have a fct "fillbyuid"  to generate a filename based on uid and template.
I am not sure about your  "./"   including of the file name.

Reply:
fixed